### PR TITLE
Add first_order

### PR DESCRIPTION
**Pull request by leeacord**
添加first order
~~FramesDataset部分似乎有个bug 正在检查~~
已修正

### CODE REVIEW

File: first_order/src/modules/dense_motion.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+        out_dict['deformation'] = deformation
+        
+        # Sec. 3.2 in the paper
+        if self.occlusion:
+            occlusion_map = fluid.layers.sigmoid(self.occlusion(prediction))
+            out_dict['occlusion_map'] = occlusion_map
+        
+        return out_dict
+
+########_inverse Test########
```
Comment (lijianshe02) [inline first_order/src/modules/dense_motion.py:179]: Running this test case when this module's name is '___main___'

File: first_order/src/modules/model.py [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+            self._layers.append(_a)
+        self.conv = dygraph.Sequential(*self._layers)
+    
+    def forward(self, x):
+        feat = self.conv_in(x)
+        out = fluid.layers.pool2d(input=self.conv(feat), pool_size=2, pool_type='max', pool_stride=2)
+        return out, feat
+
+
+class Vgg19(dygraph.Layer):
```
Comment (lijianshe02) [inline first_order/src/modules/model.py:56]: Use VGG module in paddle

File: first_order/src/modules/model.py [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+    #     # PDPD cannot use new_coordinates[..., 0]
+    #     assert len(new_coordinates.shape) == 3
+    #     grad_x = dygraph.grad(fluid.layers.reduce_sum(new_coordinates[:, :, 0]), coordinates, create_graph=True)
+    #     grad_y = dygraph.grad(fluid.layers.reduce_sum(new_coordinates[:, :, 1]), coordinates, create_graph=True)
+    #     jacobian = fluid.layers.concat([fluid.layers.unsqueeze(grad_x[0], -2), fluid.layers.unsqueeze(grad_y[0], -2)],
+    #                                    axis=-2)
+    #     return jacobian
+    
+    ## 手算一阶导的魔改版本
+    def warp_coordinates(self, coordinates_in, need_grad=False):
```
Comment (lijianshe02) [inline first_order/src/modules/model.py:207]: Does this function can solve the double gradient problem? If we use this function, will the model be trained as normal?

File: first_order/src/modules/model.py [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+    *dim_y, _, _ = y.shape
+    max_shape = np.max(np.stack([dim_x, dim_y], axis=0), axis=0)
+    if np.count_nonzero(max_shape % np.array(dim_x)) != 0 or np.count_nonzero(max_shape % np.array(dim_y)) != 0:
+        raise ValueError()
+    x_bc = fluid.layers.expand(x, (*((max_shape / np.array(dim_x)).astype(np.int32).tolist()), 1, 1)).astype('float32')
+    y_bc = fluid.layers.expand(y, (*((max_shape / np.array(dim_y)).astype(np.int32).tolist()), 1, 1)).astype('float32')
+    return x_bc, y_bc
+
+def _inverse(x):
+    """手动求逆以兼容1.8.x集群
```
Comment (lijianshe02) [inline first_order/src/modules/model.py:401]: Please write the comments in English

File: first_order/src/modules/util.py [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+#     y = torch.arange(h).type(type)
+#
+#     x = (2 * (x / (w - 1)) - 1)
+#     y = (2 * (y / (h - 1)) - 1)
+#
+#     yy = y.view(-1, 1).repeat(1, w)
+#     xx = x.view(1, -1).repeat(h, 1)
+#
+#     meshed = torch.cat([xx.unsqueeze_(2), yy.unsqueeze_(2)], 2)
+######################################################################
```
Comment (lijianshe02) [inline first_order/src/modules/util.py:70]: clean the code

File: first_order/src/train.py [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+if paddle.version.full_version == '1.8.4':
+    from paddle.fluid.dygraph.learning_rate_scheduler import MultiStepDecay
+elif paddle.version.major == '2':
+    from paddle.fluid.dygraph import MultiStepDecay
+from modules.model import GeneratorFullModel, DiscriminatorFullModel
+
+TEST_MODE = False
+if TEST_MODE:
+    logging.warning('TEST MODE: train.py')
+    # fake_input可随意指定,此处的batchsize=2
```
Comment (lijianshe02) [inline first_order/src/train.py:30]: Please write the comments in English

File: first_order/src/train.py [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+    # load start_epoch
+    if isinstance(config['ckpt_model']['start_epoch'], int):
+        start_epoch = config['ckpt_model']['start_epoch']
+    else:
+        start_epoch = 0
+    logging.info('Start Epoch is :%i' % start_epoch)
+    
+    # dataset pipeline
+    def indexGenertaor():
+        """随机生成索引序列
```
Comment (lijianshe02) [inline first_order/src/train.py:88]: Please write the comments in English

File: first_order/src/train.py [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+            dataset.buffed[i] = v.copy()
+            buf[idx] = None
+        logging.info('PreLoad Dataset: End')
+
+    if opt.mode == 'train':
+        logging.info("Start training...")
+        with dygraph.guard(plac):
+            save_dir = opt.save_dir
+            train(config, generator, discriminator, kp_detector, save_dir, dataset)
+    # elif opt.mode == 'reconstruction':
```
Comment (lijianshe02) [inline first_order/src/train.py:306]: why this mode is commented? If it can't work, please give the elaborated reason.

File: first_order/src/train.py [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+    #     print("Reconstruction...")
+    #     reconstruction(config, generator, kp_detector, opt.checkpoint, log_dir, dataset)
+    # elif opt.mode == 'animate':
```
Comment (lijianshe02) [inline first_order/src/train.py:309]: why this mode is commented? If it can't work, please give the elaborated reason.

File: first_order/README.md [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+│   ├── test
+│   └── train
+└── moving-gif
+    ├── test
+    └── train
+```
+
+## :warning:模型训练:warning:
+
+**当前模型尚未实现从头训练,且存在以下问题**
```
Comment (lijianshe02) [inline first_order/README.md:72]: The program you push on the AIStudio has displayed the training log, how that training log generated?

Comment (leeacord) [reply first_order/README.md:72]: - If you are referring to AIStudio.ipynb: In fact, this PR is based on 1.8.4 and its training function can run "normally", but the subjective visual effect of its training results is significantly worse than the official model. Thus, "The current model(script) have not realized training from scratch"

- If you are referring to my project on [aistudio](https://aistudio.baidu.com/aistudio/projectdetail/869056): The log in my aistudio is based on the 2.0rc [version](https://github.com/leeacord/first-order_PaddlePaddle/tree/work). I trained on aistudio for nearly 20 hours, but the number of steps was only 1/2.5 of the full training. I can't perform full training under the platform time limitation.

File: first_order/src/animate.py [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+        kp_new['value'] = kp_value_diff + kp_source['value']
+
+        if use_relative_jacobian:
+            jacobian_diff = np.matmul(kp_driving['jacobian'].numpy(), np.linalg.inv(kp_driving_initial['jacobian'].numpy()))
+            kp_new['jacobian'] = dygraph.to_variable(np.matmul(jacobian_diff, kp_source['jacobian'].numpy()))
+
+    return kp_new
+
+
+# def animate(config, generator, kp_detector, checkpoint, log_dir, dataset):
```
[the remark this answers is not in the document]
Comment (leeacord) [reply first_order/src/animate.py:35]: Because this PR was rewritten line-by-line from the official code, I intentionally retained the functions and structure of the official code to facilitate the reproduction of the model.
The demo.py can only process one pair of 'source_image' and 'driving_video' at once. This function is used to process many pairs at once, which is a litter bit overlap with demo.py.
I can write code to enable it if you need, or just remove it.

File: first_order/src/demo.py [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+            kp_driving = kp_detector(driving_frame)
+            kp_norm = normalize_kp(kp_source=kp_source, kp_driving=kp_driving,
+                                kp_driving_initial=kp_driving_initial, use_relative_movement=relative,
+                                use_relative_jacobian=relative, adapt_movement_scale=adapt_movement_scale)
+            out = generator(source, kp_source=kp_source, kp_driving=kp_norm)
+
+            predictions.append(np.transpose(out['prediction'].numpy(), [0, 2, 3, 1])[0])
+    return predictions
+
+# def find_best_frame(source, driving, cpu=False):
```
[the remark this answers is not in the document]
Comment (leeacord) [reply first_order/src/demo.py:79]: Because `import face_alignment` needs PyTorch. I intentionally retained the functions of the official code to facilitate the reproduction of the model in the future. Maybe it is a better choice to remove it?

File: first_order/src/train.py (hunk 9 above)
Comment (leeacord) [reply first_order/src/train.py:309]: - The demo.py can only process one pair of 'source_image' and 'driving_video' at once. Both parameters should be specified manually.
- In 'reconstruction' mode, the first frame of the input_video is used as `source_image` and the video itself is `driving_video`, which is a little bit overlaps with demo.py.
- The 'animate' mode is used to process many pairs at once, which is a litter bit overlap with demo.py.

I can write code to enable it if you need, or just remove it. The former seems to be a better choice?

File: first_order/README.md (hunk 10 above)
Comment (lijianshe02) [reply first_order/README.md:72]: You means that the 2.0rc [version](https://github.com/leeacord/first-order_PaddlePaddle/tree/work) can be trained from scratch, but the code you pushed there can not trained from scratch, am I right?

Comment (lijianshe02) [reply first_order/README.md:72]: You means that the 2.0rc version can be trained from scratch, but the code you pushed here can not trained from scratch, am I right?

Comment (leeacord) [reply first_order/README.md:72]: Because the 2.0.0rc based code has not been fully trained yet, it is difficult to say whether it can be trained from scratch or can achieve similar effects to the official model. I have noticed that the event of International Students' Day relaxes the time limitation of GPU instance in AIStudio to 168 hours. I am trying to train from scratch.
Compared to this PR (some functions have no equivalent implementation), the 2.0rc version is **obviously** closer to the official code. For the 2.0 version, the good news is that the first four significant digits of the loss value in each part are consistent with that of official code when fixing input and random numbers.

**Review by lijianshe02 — APPROVED** [review]

**Comment by huangjun12** [comment]
relative PR to PaddleGAN: https://github.com/PaddlePaddle/PaddleGAN/pull/111
